File: 55-jump-game/jump-game.cpp
class Solution {
public:
    int dp[10001];
    bool fun(vector<int>& nums,int i)
    {
        if(i>=nums.size()-1)
        {
            return true;
        }
        if(dp[i]!=-1)
        {
            return dp[i];
        }
        bool ans=false;
        for(int j=1;j<=nums[i];j++)
        {
            ans= ans || fun(nums,i+j);
        }
        return dp[i]=ans;

    }
    bool canJump(vector<int>& nums) {

        int n=nums.size(),jump=nums[0];
        for(int i=1;i<n;i++)
        {
            if(jump<i)
            {
                return false;
            }
            jump=max(jump,i+nums[i]);
        }
        return true;
    }
};

File: 55-jump-game/jump-game.java
class Solution {
    private boolean helper(int[] nums,int i,int[] visited)
    {
        if(i==nums.length-1)
        {
            return true;
        }
        boolean a= false;
        for(int j=1;j<=nums[i];j++)
        {
            if(i+j<nums.length && visited[i+j]==-1)
            {
                visited[i+j]=1;
                a= a || helper(nums,i+j,visited);
            }
        }
        return a;

    }
    public boolean canJump(int[] nums) {
        int n=nums.length;
        int[] visited = new int[n];
        Arrays.fill(visited,-1);
        visited[0]=1;
        return helper(nums,0,visited);
    }
}